hil: Add generic block device HIL

dcz-self opened the pull request:
### Pull Request Overview

This pull request adds/changes the flash HIL (https://github.com/tock/tock/pull/2248) to be more generic, and fit flash devices with or without FTL, as long as they have uniform block sizes.

Changes compared to https://github.com/tock/tock/pull/2248
- the name: block device
- most calls use block indices, not addresses. This way it's impossible to request an invalid operation.
- explicitly stated that this is for uniformly sized block devices
- erase is relaxed
- removed redundant descriptions in returns (when repeating the signature)
- documented errors on sizes
- added call to get total size

### Testing Strategy

This pull request was tested by compiling. Implementation for mx25r64 is in progress and will come later.

### TODO or Help Wanted

This pull request still needs reviews. In particular, I'm not sure if a future HIL suitable for flash should rather go in the more generic, or more specialized direction (in particular, non-uniform block/page/sector sizes).

I'm also not sure if it's a good idea to read/write/erase multiple blocks in one go. Perhaps call back multiple times, once per block? Or limit to one block per request?

### Documentation Updated

- [ ] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make prepush`.

dcz-self commented:
I added a basic implementation and made a simple test on real HW in https://github.com/dcz-self/tock/commits/flashil .

bradjc commented on kernel/src/hil/block_storage.rs:
```diff
@@ -0,0 +1,268 @@
+    ///
+    /// If `region` exceeds the size of the device, returns `ErrorCode::INVAL`.
+    /// On success, triggers `Client::erase_complete` once.
+    fn erase(&self, region: &Region<E>) -> Result<(), ErrorCode>;
```
This description seems like a different operation than "erase". I think this should either promise to clear data or be called `prepare_write()` (or something similar).

bradjc commented on kernel/src/hil/block_storage.rs:
```diff
@@ -0,0 +1,268 @@
+    fn erase(&self, region: &Region<E>) -> Result<(), ErrorCode>;
+
+    /// Returns the size of the device in bytes.
+    fn get_size(&self) -> Result<u64, ErrorCode>;
```
How could this fail?

dcz-self replied:
I think this could be reasonably expected to ask the device to identify itself, and fail if the size can't be determined.

On the other hand, perhaps it could be made infallible. The other methods are already supposed to fail if bounds are exceeded, and it doesn't make a lot of sense to not know how big a device is from the init.

I could use some feedback on this.

bradjc replied:
Well if this leads to a device query then it would have to be asynchronous, or the device query would have to happen in advance which complicates the implementation.

But in general, are there hardware designs which wouldn't know the size of their block storage? It seems like it's a reasonable assumption for an embedded device to make that it knows the size of its storage.

dcz-self replied:
I removed the possibility to have an error completely. The data manipulation needs to know the device size anyway to be able to error out early anyway.

bradjc commented on kernel/src/hil/block_storage.rs:
```diff
@@ -0,0 +1,268 @@
+    /// Block read complete.
+    ///
+    /// This will be called when the read operation is complete.
+    fn read_complete(&self, read_buffer: &'static mut [u8], ret: Result<(), ErrorCode>);
```
What are the errors?

dcz-self replied:
I left it for now. I presume that any virtualization needs to store the read offset anyway, so storing read size alongside it wouldn't be a problem. Or am I mistaken?

phil-levis replied:
Given both BlockIndex and AddressRange as borrowed with an anonymous lifetime, I think this is OK. It does necessitate that the caller keeps that state. I see benefits and drawbacks in both approaches (having the caller keep the state, or having the callee pass it back).

bradjc commented on kernel/src/hil/block_storage.rs:
```diff
@@ -0,0 +1,268 @@
+    /// starting at the chosen `start_block`.
+    ///
+    /// `ErrorCode::INVAL` will be returned if
+    /// - `range` exceeds the end of the device, or
```
No `range`

bradjc commented on kernel/src/hil/block_storage.rs:
```diff
@@ -0,0 +1,268 @@
+    /// - `range` exceeds the end of the device, or
+    /// - `buf` is shorter than `region`.
+    ///
+    /// This function SHALL NOT prepare the block for writing first.
```
```suggestion
    /// This function SHALL NOT prepare the region for writing first.
```
? Or "blocks in region" maybe.

phil-levis replied:
I'm wary of this SHALL NOT, because it's an implementation decision. I think you want to say that this method does not any claims of preparation, and that a client that needs a clean block MUST call `prepare_write` first. That leaves open the possibility of multiple writes, while also allowing implementations that defer preparation to this actual method call (e.g., by storing a "prepare first" bit).

bradjc commented on kernel/src/hil/block_storage.rs:
```diff
@@ -0,0 +1,268 @@
+        buf: &'static mut [u8],
+    ) -> Result<(), (ErrorCode, &'static mut [u8])>;
+
+    /// Read data from blocks starting at `block`, and into the buffer.
```
region

bradjc commented on kernel/src/hil/block_storage.rs:
```diff
@@ -0,0 +1,268 @@
+                    range.start_address % S as u64 + range.length_bytes as u64,
+                    S as u64,
+                ) as u32
+                // CAUTION: div_ceil is nightly
```
what does this refer to?

dcz-self replied:
It's the `div_ceil` function which is needed here: https://doc.rust-lang.org/std/primitive.u64.html#method.div_ceil . This is also what the FIXME is referring to; I haven't found the right incantation to do the division yet.

bradjc replied:
Ok makes sense, I was just expecting to see a feature.

bradjc reviewed (commented):
I'm not the best to judge if this meets the various needs for an alternative flash hil, but I made some comments on the HIL and documentation.

dcz-self commented on kernel/src/hil/block_storage.rs:
```diff
@@ -0,0 +1,268 @@
+
+/// An index to a block within device composed of `S`-sized blocks.
+#[derive(Clone, Copy)]
+pub struct BlockIndex<const S: u32>(pub u32);
```
I think `Index` works to disambiguate that it's an indirection, and not e.g. a copy of the data.

dcz-self commented:
Thanks for reviewing, especially proofreading. I changed the names a few times and that left stale traces.

After implementing a userspace driver, I'm more and more tempted to drop the ability to affect multiple blocks in a row. Supporting multiple blocks makes the internal capsule logic more complex, and it's not clear to me what the API should do about buffer sizes. Unless there's good reason to keep it in the kernel, I would rather do that in the userspace.

That being said, my file system experience is limited. Perhaps it makes sense from the performance POV. Feedback welcome.

dcz-self commented on kernel/src/hil/block_storage.rs:
```diff
@@ -0,0 +1,268 @@
+    ///
+    /// On success, triggers `Client::read_complete` once.
+    /// On failure returns a `ErrorCode` and the buffer passed in.
+    /// If `ErrorCode::NOSUPPORT` is returned then `read_block`
```
Split into another trait.

phil-levis commented on kernel/src/hil/block_storage.rs:
```diff
@@ -0,0 +1,238 @@
+    /// Makes a region ready for writing.
+    ///
+    /// This corresponds roughly to the erase operation on raw flash devices,
+    /// but may also do nothing on devices where erasure is not necessary.
```
This suggests that this API does not support the possibility of multiple writes (e.g., on flash)?

I think that's OK, but this should be explicit.

dcz-self replied:
I think you're reading this right: regions on flash are not ready for writing unless they get erased first. Without this method, we can't support flash.

This particular passage says that if you have 3D Xpoint memory or a plain old HDD, this may be implemented as a no-op.

phil-levis replied:
This starts to get a little weird. It means that a client on top of an HDD can read discarded data, while a client on top of flash can't. This then exposes which one the client is on top of, such that a client sitting on top of a HDD will do things that then won't work if you replace the underlying block store. This means that API is no longer transparent and isn't an interchangeable API.

dcz-self replied:
Perhaps I didn't make it clear, but my intention here is that the result of reads of discarded blocks are undefined. The note that discard may be a no-op is a hint to the driver author, rather than an API guarantee.

By assuming SSD semantics of required discard, the API remains interchangeable (if SSD/HDD interchangeability is what you refer to).

Would it be clearer with the following wording?

> This corresponds roughly to the erase operation on raw flash devices.
> An implementation may internally decide to do nothing when irrelevant,
> e.g. on devices where erasure is not necessary.

phil-levis replied:
The issue with "undefined" is that a client which assumes a particular implementation is then bound to it; it's no longer an interchangeable interface. The client uses a `BlockStorage`, but it has to be a specific one or it won't work. I.e., a client built on top of an HDD or NVRAM device could read after discarding, but then if you put it on top of an SSD it won't work.

phil-levis replied:
> This corresponds roughly to the erase operation on raw flash devices.
> An implementation may internally decide to do nothing when irrelevant,
> e.g. on devices where erasure is not necessary.

The issue is what the client will do.

dcz-self replied:
I see "undefined" more as "forbidden, but not enforced", as opposed to "implementation-defined". If a client decides to rely on something that's outside of the API contract, then it's the client's problem. Within the constraint that a client only uses allowed operations, it won't be able to distinguish between a HDD or SDD.

I admit it's not perfect to leave an undefined operation, but I think this is a local optimum for the useablity of this API. If we tried to eliminate that undefinedness, I expect we'd have to come up with something a lot more complicated. Perhaps we want to in the future, but I think this proposal is a good stepping stone to open up the familiar block storage functionality at all.

phil-levis commented on kernel/src/hil/block_storage.rs:
```diff
@@ -0,0 +1,238 @@
+    /// as part of a prepare block.
+    /// Multiple writes to the same block are possible in this trait,
+    /// but the result is dependent on the nature of the underlying hardware,
+    /// and this interface does not provide a way to discover that.
```
I am wary of this SHOULD NOT. It means that someone can use an instance of this trait and make implementation-specific or hardware-specific assumptions, which then break on another implementation. This should either not be general (distinguish write models), or a MUST.

dcz-self replied:
Sure. I wanted to accommodate NOR flash devices (they allow multiple writes without erasing), but it's probably better to not touch it until there's an actual need.

phil-levis replied:
I believe some NAND devices do as well (although fewer writes).

alistair23 replied:
We should accommodate multiple writes without an erase. It's a pretty common use case, one that [TicKV](https://github.com/tock/tock/tree/master/libraries/tickv) and [Yaffs1](https://yaffs.net/documents/how-yaffs-works) both rely on for example

dcz-self replied:
I want to address this in a future revision. Currently we managed to work out a compromise based on the agreement that the targeted devices are basic, and the user does not need to know about how many writes they can receive.
Besides, the current version Works For Me™ ;) That is, it's already covering some useful ground.

phil-levis commented on kernel/src/hil/block_storage.rs:
```diff
@@ -0,0 +1,238 @@
+    ///
+    /// `ErrorCode::INVAL` will be returned if
+    /// - `region` exceeds the end of the device, or
+    /// - `buf` is shorter than `region`.
```
Distinguish these two cases with SIZE?

phil-levis reviewed (changes requested):
Overall, I think this looks pretty good. By just trying to do a simpler, general block interface it avoids all of the difficult complexities of flash. However, there's one detail I'm worried about. It's often the case that the erase unit for flash is larger than the write unit. In that case, `prepare_write` can have side effects. Or is the intention that this API would just make the write unit as large as the erase unit?

Finally, I would recommend separating the read and write APIs, then joining them in a composite trait. With this API, you can't give a client just read access or just write access. Please take a look at https://github.com/tock/tock/blob/master/doc/reference/trd2-hil-design.md#rule-8-use-fine-grained-traits-that-separate-different-use-cases

dcz-self commented:
I originally had the "prepare_write" method called "erase", but that was not clear enough. "discard" is used in other contexts, and I think it's clearer that its block size is not the same as the write size (especially after I rename the "erase block size" to "discard block size"). Would that be clearer?

phil-levis commented:
> I originally had the "prepare_write" method called "erase", but that was not clear enough. "discard" is used in other contexts, and I think it's clearer that its block size is not the same as the write size (especially after I rename the "erase block size" to "discard block size"). Would that be clearer?

I think, for this API, separating the discard and write sizes is a big change. There are a lot of weird cases. In particular, what happens if you read after discard? With this API, a device using a block device (e.g., HDD, a RAM device, or a NVRAM device) has to manage the complexity of discard sizes potentially being larger than write sizes, even though discard is a no-op. I.e., if it wants to change one block of a page, it has to copy out the other blocks first, discard the page, then re-write the blocks. I thought you were going in the opposite direction, that for this simple API the write unit is the same as the discard unit. The challenge here, though, is that the erase unit for large NAND is often 256kB.

There are NAND flashes that allow multiple writes to a single block. This is sometimes used by flash-specific storage abstractions, e.g., to maintain free lists as bitmasks, etc. I don't think this API needs to represent this. 

There's been I think now 3 attempts to write a better API for flash, that exposes all of its pecularities. It's tough. One effort didn't even make it to be a PR, here are the other two:

https://github.com/tock/tock/pull/2249
https://github.com/tock/tock/pull/2248
https://github.com/tock/tock/pull/1457

dcz-self commented:
> what happens if you read after discard?

As mentioned in the other comment, I think the only sane way out is to have the consumer suffer the consequences. It's infeasible to try to react to bad reads, because that would require tracking erased blocks, which I think is out of scope of a low-level layer like this.

> With this API, a device using a block device (e.g., HDD, a RAM device, or a NVRAM device) has to manage the complexity of discard sizes potentially being larger than write sizes, even though discard is a no-op.

The block sizes are set by the driver, so if the driver determines that discard is a no-op, it is free to define all block sizes as the same value.

phil-levis commented:
> > what happens if you read after discard?
> 
> As mentioned in the other comment, I think the only sane way out is to have the consumer suffer the consequences. It's infeasible to try to react to bad reads, because that would require tracking erased blocks, which I think is out of scope of a low-level layer like this.

I think we're talking at cross purposes here. You're discussing what an implementation could do, given an ambiguous specification that allows multiple behaviors. I'm pointing out this means that a client written for one implementation might not work with others because it assumes those behaviors. The API doesn't fully encapsulate the abstraction and create a layered system; information about the lower layer can leak into the upper layer. In that case, we should just define a block API and a flash API.

dcz-self commented:
> a client written for one implementation might not work with others because it assumes those behaviors

Right, I'm fine with letting the client live in a hell of its own making by choosing to adhere to the implementation rather than the contract ;)

More seriously, like I wrote before, I don't see any more practical option at the moment. This API is primarily meant to cover devices that require erasure before writing. I don't currently see a way to define an API that enforces erasure before writing without revealing underlying implementation details, so splitting this proposal into 2 APIs doesn't bring any benefits that I can see.

dcz-self commented:
To add on to the previous comment – if we wanted to be explicit about the underlying implementation, we'd have to define what happens to the erased blocks on NAND and NOR devices, and we'd have to define how many writes are allowed. The first one means we can only support devices which set bits to 1 on erase (reasonable enough), and those which may be rewritten at most once (nrf52 supports writing twice, so would not be supported). Because if a chip supports two rewrites, a client could try to rely on this property of the implementation, and we would have to split out an API supporting 2 rewrites as a separate driver.

I'm not sure it makes sense to try to avoid all such problems explicitly, strictly from the practical point of view. I'd rather use this HIL as a stepping stone to implement a file system, and recommend an actual file system in the long term, after marking the block driver "unsafe" in some way.

dcz-self commented:
Having deliberated on this some more, I know how to force an erase before a write: by returning a "ticket" from an erase operation, and consuming the ticket as a requirement for the write operation.

I wasn't able to invent any mechanism to prevent a read of erased data. But I don't think we can completely prevent that without rolling out an entire file system: it will always be possible to read erased blocks right after the driver was loaded, as erasure bitmaps are not preserved. This means that the specifics of the medium will always be accessible to the client, so it remains for us to choose whether we want to mandate particular implementations in this HIL or not.

phil-levis commented:
> Having deliberated on this some more, I know how to force an erase before a write: by returning a "ticket" from an erase operation, and consuming the ticket as a requirement for the write operation.

Excellent!

> I wasn't able to invent any mechanism to prevent a read of erased data. But I don't think we can completely prevent that without rolling out an entire file system: it will always be possible to read erased blocks right after the driver was loaded, as erasure bitmaps are not preserved. This means that the specifics of the medium will always be accessible to the client, so it remains for us to choose whether we want to mandate particular implementations in this HIL or not.

Given that you need to separate the read and write paths, and the fact that the actual timing of erases is implementation-dependent (they could be deferred, occur on the next write, etc), that's OK.

How does this work if the erase and write units are different? Does an erase give you a set of write tokens?

dcz-self commented:
> How does this work if the erase and write units are different? Does an erase give you a set of write tokens?

Correct.

The downside of this approach is that the driver now needs to be able to allocate memory. Handing out opaque tokens won't cut it. Consider:

```
let t1 = discard(0);
let t2 = discard(0);
write(t1, data1);
write(t2, data2); // 2 writes, we're doomed
```

To prevent the double-write, the driver could prevent handing out the same token twice. It must therefore keep track of all active tokens.

Option 1: the driver allocates the memory itself. The driver must either utilize allocation or have a limit on the number of tokens in flight.
Option 2: the driver uses memory provided by the client. The API would become sth like `discard(block: u32, space: &'static mut Space) -> Token` - by similarity to the current way of sending buffers to SPI devices. The buffer would be returned in `write(Token) -> &'static mut Space`, where the client could get rid of the memory. The problem with this approach is that it's now impossible to use the HIL if the number of tokens in flight is unknown, like in a userspace driver.

I think all options here suck, although the version which limits the number of active tokens seems simple enough to try out. It might even survive a meeting with reality if the file system built on top is conservative about the number of used tokens. Ultimately though, what we're trying to do here is to build an occupancy bitmap, which, as I wrote before, I consider overkill for this application as it belongs in the file system layer.

hudson-ayers commented:
> > > what happens if you read after discard?
> 
> > 
> 
> > As mentioned in the other comment, I think the only sane way out is to have the consumer suffer the consequences. It's infeasible to try to react to bad reads, because that would require tracking erased blocks, which I think is out of scope of a low-level layer like this.
> 
> 
> 
> I think we're talking at cross purposes here. You're discussing what an implementation could do, given an ambiguous specification that allows multiple behaviors. I'm pointing out this means that a client written for one implementation might not work with others because it assumes those behaviors. The API doesn't fully encapsulate the abstraction and create a layered system; information about the lower layer can leak into the upper layer. In that case, we should just define a block API and a flash API.

"Using the API in an unsupported way is not allowed to leak any information about the lower layer" seems like a higher isolation bar than we generally enforce for Tock HILs, and I am not sure that the benefits (preventing someone writing a driver that uses the API in a way forbidden by the docs but happens to work on only some platforms) of enforcing it here are worth the costs (significant implementation complexity, and likely size and runtime degradation).

If the concern is that someone might rely on erase to clear sensitive data, the HIL should just clearly document that you must overwrite sensitive data if you want to guarantee it is not accessible in the future.

phil-levis commented:
> > How does this work if the erase and write units are different? Does an erase give you a set of write tokens?
> 
> Correct.
> 
> The downside of this approach is that the driver now needs to be able to allocate memory. Handing out opaque tokens won't cut it. Consider:
> 
> ```
> let t1 = discard(0);
> let t2 = discard(0);
> write(t1, data1);
> write(t2, data2); // 2 writes, we're doomed
> ```

Yeah, without keeping per-page metadata this doesn't work; that's why FTLs are like they are.

What size flashes are you thinking of? A 16MB flash with 4kB pages, for example, can have a "erased" bitmask of 512 bytes.

It's my experience that you want to treat FTL'd and non-FTL'd chips differently. You want to use different storage abstractions. E.g., without an FTL you need to do wear-levelling or you can easily lose blocks. Unless you expect everything using this layer to use logging approaches and just be inefficient on an FTL device?

phil-levis commented:
> seems like a higher isolation bar than we generally enforce for Tock HILs, and I am not sure that the benefits (preventing someone writing a driver that uses the API in a way forbidden by the docs but happens to work on only some platforms) of enforcing it here are worth the costs (significant implementation complexity, and likely size and runtime degradation).

*shrug* It'ss just good software design. Otherwise you get hidden dependencies and pieces are not composable. The cost you pay here (which I mention in a separate comment above) is that you have to assume there isn't an FTL, which means all of your storage approaches need to be log-structured or you can quickly (e.g., 100 erases) lose blocks. So providing this API on erase-free devices (HDD or FTL'd flash) imposes a large complexity overhead.

If we want to say that this API is really intended for raw flash, but of course you can put it on top of an FTL device and pay a huge penalty, I'd be OK with that. I'd expect we'd have a different API for a write-in-place device, and a trivial shim (ignores erase operations). It does get a little weird in that many flash storage abstractions assume the erase behavior of flash (e.g., for bitmasks). It would mean that if you want to be sure about erase behavior your actually need to erase, write, erase.

dcz-self commented:
I don't like the RAM occupancy map, because it imposes an extra resource cost on all users. Twofold: if the client tries to be lean by storing occupancy on the flash device itself, it doesn't need any of this in the RAM. Secondly, when storing a bitmap of erased blocks, every block erased before initialization will be erased again before writing, ~halving the lifetime of the device in some scenarios.

I think there might be a place for such a driver. But still, implementing an erasure bitmap in RAM will not completely prevent the client from figuring out what the lower layer is: all the client needs to do is to erase a block, wait until the driver is reinitialized, and then read it.

This HIL is not suitable for FTL-naive clients. The API makes it possible to use any kind of device (FTL or not. It's impossible to prevent a write-many-times device from exposing an erase-write interface), but by requiring erasure, it necessarily requires clients to do the management themselves. Your observations on HDDs are correct. I mostly left them in because why not - HDDs are a generalization of flash, and I want my code work on them too, even when it's not the best match.

> It would mean that if you want to be sure about erase behavior your actually need to erase, write, erase.

Could you provide an example of when that would be needed?

phil-levis commented:
> I don't like the RAM occupancy map, because it imposes an extra resource cost on all users. Twofold: if the client tries to be lean by storing occupancy on the flash device itself, it doesn't need any of this in the RAM. Secondly, when storing a bitmap of erased blocks, every block erased before initialization will be erased again before writing, ~halving the lifetime of the device in some scenarios.

I agree, I don't either. 

> This HIL is not suitable for FTL-naive clients. The API makes it possible to use any kind of device (FTL or not. It's impossible to prevent a write-many-times device from exposing an erase-write interface), but by requiring erasure, it necessarily requires clients to do the management themselves. Your observations on HDDs are correct. I mostly left them in because why not - HDDs are a generalization of flash, and I want my code work on them too, even when it's not the best match.

The first sentence of the PR says: *and fit flash devices with or without FTL,*

I think your comments are a bit contradictory? FTL-native devices act like standard block devices, allowing in-place writes: the FTL allows a flash device to act like an HDD. So if your code works on an HDD, it'll work on an FTL flash device. 

> > It would mean that if you want to be sure about erase behavior your actually need to erase, write, erase.
> 
> Could you provide an example of when that would be needed?

A storage layer is keeping a bitmask, denoting the use of a particular resource (blocks, etc.). It calls erase and expects the bits to be cleared (let's suppose it has a clean way to know whether `1` is cleared or `0` is depending on medium). It later reads the bits to see which blocks are in use. If it's using an implementation that just discards erases it will see the old values. So the storage layer actually needs to erase-write to be sure the bits are cleared, then erase so it can start writing them again. So erase-write-erase. Alternatively, implementations of this trait on HDD/FTL/NVRAM would need to write to the block to clear it on an erase operation. Put another way, an erase is a type of write, just a very specific one.

dcz-self commented:
> The first sentence of the PR says: and fit flash devices with or without FTL,
>
> I think your comments are a bit contradictory? FTL-native devices

Note that I did not write "na**t**ive devices", but "naive *clients*". The client has to expect an FTL, the device is unconstrained.

Thanks for explaining the erase-write-erase. Yes, that's a problem after each reset of the state of the storage layer, so best to minimize those IMO.

Where do we stand on it? As far as I can tell:

- rename the "erase" functionality to use the word "discard",
- split out a read-only trait,

is that all to reach a level of usefulness sufficient for inclusion?

phil-levis commented:
> > The first sentence of the PR says: and fit flash devices with or without FTL,
> > I think your comments are a bit contradictory? FTL-native devices
> 
> Note that I did not write "na**t**ive devices", but "naive _clients_". The client has to expect an FTL, the device is unconstrained.

I'm confused. Do you mean that clients of this interface expect that it has an FTL, but implementations may be on top of raw flash and implement the FTL? If so, why is there a discard operation? Flash translation layers abstract away erases and expose a standard block-write interface, just like an HDD.

> Thanks for explaining the erase-write-erase. Yes, that's a problem after each reset of the state of the storage layer, so best to minimize those IMO.

What do you mean? It's a logical error that can lead to undefined results -- how does the system handle it? Minimization isn't sufficient; what's needed is correctness.
 
> Where do we stand on it? As far as I can tell:
> 
>     * rename the "erase" functionality to use the word "discard",
> 
>     * split out a read-only trait,
> 
> 
> is that all to reach a level of usefulness sufficient for inclusion?

It needs one of three changes:
  - Be defined only for raw flash devices (which have a clear behavior for discard),
  - Be defined for FTL and other block devices and remove the discard operation,
  - Define the behavior of discard (what happens if you read a block after discarding its page), noting that for FTL and other block devices, discard means "write all 0s" or "write all 1s".

dcz-self commented:
> Do you mean that clients of this interface expect that it has an FTL, but implementations may be on top of raw flash and implement the FTL?

The opposite: clients must be able to deal without an FTL. Implementations may or may not have an FTL.

> What do you mean? It's a logical error that can lead to undefined results -- how does the system handle it? Minimization isn't sufficient; what's needed is correctness.

I mean that when the layer above the device tracking erase state is uninitialized, it knows nothing about the state of the device. It can't be fixed by definition. By minimization I meant: better not keep the state in volatile storage.

> Be defined only for raw flash devices (which have a clear behavior for discard),

Ok, this seems reasonable, and if it needs extending to non-flash, that's also a no-op. I'll do it.

dcz-self commented:
Sure, I get your point. But

> If the implementation of the trait does have an FTL, then the client using it can treat it like a standard block device and perform arbitrary numbers in-place writes on a single block.

Even if the FTL is not present, the client can treat it like a block device. But by doing this, it's going to break the API contract, regardless of whether there's a FTL, or not. Putting a comment in the docs to the effect of "underlying device does not contain FTL" will have no bearing on how badly the API can be abused, and I meant that removing that comment is a no-op :)

dcz-self commented:
Thanks for all the reviews. I've done the requested changes:
- separated read and write traits
- removed comments trying to make this HIL too general
- noted the precise semantics of discard: obligatory, not enforced, leaves data undefined

phil-levis commented:
Two changes needed as part of the splitting into read and write traits;
  - You need to [split the `Client` traits and add `set_client` methods to the separate traits for each type of client](https://github.com/tock/tock/blob/master/doc/reference/trd2-hil-design.md#rule-12-traits-that-can-trigger-callbacks-should-have-a-set_client-method). 
  - Please need to add a [composite trait (with automatic implementation) that combines Read and Write](https://github.com/tock/tock/blob/68cf7940a745e4bba2fe4a1409683443a8348302/kernel/src/hil/uart.rs#L62).

phil-levis commented:
Other than that, looks good!

phil-levis commented:
I am not sure if we want to merge this before the 2.1 release -- I am guessing no. It would be good to write a draft TRD describing the design of the API, but merging shouldn't block on this.

dcz-self commented:
I've added requested changes: Client is now split, and there is a composite Client and Storage.

While adjusting the first driver, I noticed that some convenience methods were missing, so there's now a From<BlockIndex> for AddressRange.

dcz-self commented:
After I am finished with the driver, I'll fork the syscall interface and write a TRD as well, submitting as I go.

dcz-self commented:
Minor changes: typos and rewording.

dcz-self commented:
I've drafted a TRD https://github.com/tock/tock/pull/3157

alistair23 commented on kernel/src/hil/block_storage.rs:
```diff
@@ -0,0 +1,247 @@
+    /// The user of this function MUST ensure that the relevant block
+    /// has been successfully discarded first (see `discard`).
+    ///
+    /// Once a byte has been written as part of a write block,
+    /// it MUST NOT be written again until it's discarded
+    /// as part of a discard block.
```
This is very restrictive and doesn't match a lot of flash use cases

dcz-self replied:
Covering all use cases is not the goal at this stage, see discussion leading up to https://github.com/tock/tock/pull/2993#issuecomment-1207910434

alistair23 replied:
I would say allowing multiple writes doesn't mean covering all use cases, just a large number of them

dcz-self replied:
I agree, but I still think it's better to tackle this as a separate unit of work. Incremental progress seems to be working out better in Tock than general solutions in my experience.

phil-levis replied:
I agree -- all of previous attempts to write a powerful but general flash API have failed because of all of the tricky edge cases: the APIs worked onlu for the particular devices/use cases the author cared about. We can go down that road again, but this shouldn't block having a simpler API for simpler use cases.

alistair23 commented on kernel/src/hil/block_storage.rs:
```diff
@@ -0,0 +1,247 @@
+/// Specifies a storage area with byte granularity.
+pub struct AddressRange {
+    /// Offset from the beginning of the storage device.
+    pub start_address: u64,
```
`usize` for address?

dcz-self replied:
The storage addresses are not in the CPU memory space, so I opted for making the size independent of CPU architecture.

twilfredo commented on kernel/src/hil/block_storage.rs:
```diff
@@ -0,0 +1,268 @@
+
+/// Divide, round up
+fn div_ceil(a: u64, b: u64) -> u64 {
+    a / b // FIXME: trololo
```
Is this what you are after here `(a + b-1) / b`?

phil-levis commented:
Overall, I think this is in good shape. My one major comment is that since the API is now designed to encapsulate raw flash storage, rather than block devices generally, it should probably not be called "block storage". Instead, should this replace `hil::flash`? I am guessing no due to @alistair23's concerns. However, it might make sense to put it in `hil::flash` as something like "BlockFlash" and explicitly distinguish the two. Otherwise someone looking through the HILs might be confused on which one to use when and why. E.g., the current `hil::flash` only allow reading/writing of pages, while this new API
allows block-level reads and writes.

A "discard block" should be called a "page" (standard flash terminology).

I suspect that we will at some point in the future want to break out discard from write (e.g., so a client can be given the ability to write but not delete data), but I think we can leave them coupled for now.

dcz-self commented:
I still think that supporting discard doesn't automatically make this a flash-only driver. Discard is also seen in VM images, for example. Naming the API after flash would make less sense than naming an API without discard after HDDs: no discard makes it unsuitable for flash, presence of discard does not make it unsuitable for HDDs (even if today we ban it by decree).

On the other hand, it's a minor thing, and given the embedded focus, it's probably not going to be used on anything but flash anyway.

alistair23 commented:
This shouldn't replace the flash HIL as the lack of multi-write support is very limiting. Adding it as a `block_device` HIL sounds fine to me

phil-levis commented:
> This shouldn't replace the flash HIL as the lack of multi-write support is very limiting. Adding it as a `block_device` HIL sounds fine to me

It definitely shouldn't replace. The question is whether is it is a flash HIL (as the TRD says it is) or a HIL for more general durable storage.

"Block device" is a technical term, from the original AT&T UNIX and later POSIX (character devices and and block devices). We should not reuse it unless the abstraction is extremely similar and it behaves like a block device, which this does not. If the API described block devices, it should be called `block_device`, but since it doesn't (discard, page vs. block), it shouldn't. E.g., https://docs.oracle.com/cd/E19683-01/806-5222/block-34861/index.html

bradjc commented on kernel/src/hil/block_storage.rs:
```diff
@@ -0,0 +1,247 @@
+    /// Returns `ErrorCode::BUSY` when another operation is in progress.
+    ///
+    /// On success, triggers `WriteableClient::discard_complete` once.
+    fn discard(&self, region: &BlockIndex<D>) -> Result<(), ErrorCode>;
```
The unit of `D` here is bytes?

Doesn't that conflict with `BlockIndex<S>` where `S` is in blocks?

What is a region?

I think we need a glossary in the top comment. We have blocks, write blocks, discard blocks, regions, `S`, `D`, `W`, `R`, at least.

bradjc commented on kernel/src/hil/block_storage.rs:
```diff
@@ -0,0 +1,247 @@
+use crate::ErrorCode;
+use core::ops::Add;
+
+/// An index to a block within device composed of `S`-sized blocks.
+/// Stores the number of blocks from the start of the device.
+#[derive(Clone, Copy, PartialEq, Eq)]
+pub struct BlockIndex<const S: usize>(pub u32);
```
The unit of `S` is bytes?

bradjc commented on kernel/src/hil/block_storage.rs:
```diff
@@ -0,0 +1,247 @@
+}
+
+/// Devices which can read arbitrary byte-indexed ranges.
+pub trait ReadRange {
```
This trait is for a device?

Also, it seems out of place to have byte level reads in a block storage driver.

dcz-self replied:
I am not particularly attached to it, but I've noticed that flash devices include byte addressing mode for reads. Should I get rid of it?

bradjc replied:
What about having `ReadableStorageBlocks` and `ReadableStorageBytes`. Then drivers can implement whatever they support.

bradjc commented on kernel/src/hil/block_storage.rs:
```diff
@@ -0,0 +1,247 @@
+/// Specifies a storage area with byte granularity.
+pub struct AddressRange {
+    /// Offset from the beginning of the storage device.
+    pub start_address: u64,
```
This is in blocks or bytes?

dcz-self replied:
Corrected docstring to bytes.

bradjc reviewed (commented):
It seems like we 1) really want something like this, and 2) have a hard time agreeing on what that something is. I think the way forward is to, as discussed, not solve every problem, and not over-promise in the API. However, I think the tradeoff is we need more clarity on what this HIL actually does so that a future reader can understand if it will work for their use case or not. Right now the assumptions and intent aren't clear.

Also I get confused about whether this is for flash, or this isn't for flash. I think we either say this is designed explicitly for some subset of flash chips and make it clear that the particulars of the API are based on flash limitations, and name it as such, or we don't, and we remove the subtle references to flash.

dcz-self commented on kernel/src/hil/block_storage.rs:
```diff
@@ -0,0 +1,247 @@
+    /// Offset from the beginning of the storage device.
+    pub start_address: u64,
+    /// Length of the range.
+    pub length_bytes: u32,
```
The name indicates bytes.

dcz-self commented:
> The question is whether is it is a flash HIL (as the TRD says it is) or a HIL for more general durable storage.

I'm fine to limit the intentions for this version to raw flash only, for the sake of limiting the scope of discussion. I will definitely not hold myself back in the future when I get a chip with an FTL. I will just reuse this, and try to apply the lessons to refine the API.

After reading http://www.linux-mtd.infradead.org/faq/general.html#L_mtd_vs_hdd , I agree that this is not falling under the strict definition of "block device", although I have no proposals what to call the superset of block device and raw flash. Not that I insist on naming the HIL after that name.

I think the docs clearly indicate flash in this revision. I am happy to rename it to whatever is deemed appropriate.

dcz-self reviewed (commented):
I updated the proposal with some new feedback.

bradjc commented on kernel/src/hil/block_storage.rs:
```diff
@@ -0,0 +1,248 @@
+
+impl<const B: usize> BlockIndex<B> {
+    /// Returns the index that contains the address.
+    pub fn new_containing(byte_address: u64) -> Self {
```
Why is this `u64`? Tock only supports 32 bit platforms.

dcz-self replied:
This is to support storage devices with more than 4GiB capacity.

bradjc commented on kernel/src/hil/block_storage.rs:
```diff
@@ -0,0 +1,248 @@
+    }
+}
+
+impl<const B: usize> From<BlockIndex<B>> for u64 {
+    fn from(index: BlockIndex<B>) -> Self {
+        index.0 as u64 * B as u64
+    }
+}
```
Why do we need this? It's not clear to me that converting a blockindex to a u64 would give the address of the start of the block. Why not something explicit like `BlockIndex.get_start_address()`?

phil-levis commented:
Comparing this with https://github.com/tock/tock/pull/2248, I think there are relevant differences based on different use cases. Until (unless?) we come up with the "one true flash HIL", I support having multiple HILs that can support different use cases. There is of course the challenge that you can't easily compose the use cases, but you also can't if you only support one of the HILs. :) I'd describe the differences as these:
  - Current API - Supports only page-based access, indicated by a page index. Read, write and erase are all in the same trait.
  - #2248 - Supports only page-based access, uses a `Page` object for verifying valid addresses. No implementations of the trait? Read, write and erase are all in the same trait. Can express that write and erase page sizes can be different (technically, should be block and page).
  - #2993 (this) - Separates trait into (write, erase) and read. Supports arbitrary address range reads, page-sized erases and block-sized writes. 

Here are my suggested changes (open to discussion):
  - [ ] Move these traits into `hil::flash`
  - [ ] Changes names to `Read`, `Modify`, `ReadClient`, and `ModifyClient`
  - [ ] Add `set_client()` methods to `Read` and `Modify`. 
  - [ ] Merge `ReadableStorage`/`Read` and `ReadRange` into a single trait, since they trigger the same callback. Alternatively, make `ReadRange` an extension of `Read`. Otherwise there is a problematic discontinuity, where the runtime behavior of one trait depends on the calls onto another (e.g., I assume that if a client calls `read` then `read_range` the second call will fail with BUSY?). 
  - [ ] Break `BlockIndex` into `BlockIndex` and `PageIndex` rather than just distinguishing on the size -- they really are different abstractions and following flash terminology is helpful
  - [ ] In the `Ok` value of the `read_complete` event, include a `usize` to indicate how much was read. For `read` this will always be R, but for `read_range` it could be something else. This provides a simple way for a client to know exactly how much of the data in `buf` contains read data, rather than forcing the client to track this in internal state.
  - [ ] Change the `flash` documentation to note that there are two APIs, how they differ, and why.

bradjc commented:
Sounds good to me, but I think we need a name for the new traits, so we don't have confusion about the difference between the current `hil::flash::Flash` and what would be `hil::flash::Read`. Also, since we aren't trying to solve all flash use cases, we should save some namespace for future interfaces.

phil-levis commented:
> Sounds good to me, but I think we need a name for the new traits, so we don't have confusion about the difference between the current `hil::flash::Flash` and what would be `hil::flash::Read`. Also, since we aren't trying to solve all flash use cases, we should save some namespace for future interfaces.

I don't see a good way to assign names that are simultaneously descriptive for the different APIs and are succinct. So I see two options:
  1. Use proper nouns: call them something like "flash::ainu::" and "flash::beagle::" Authors can decide on the names.
  2. Keep the names as is and have the documentation clearly articulate that certain sets of traits go together and what they are.

Or we can try to for something descriptive, like `flash::PageFlash` and `flash::PageBlockRead`, `flash::PageBlockModify`.

alevy commented:
This seems worth reviving to me (I have the hardware this was designed for, so I'm biased). This seems to have been very close and the dependent PRs on the surface seem fairly easy to approve other than the HIL discussion.

If @dcz-self has moved on, I'm willing to "take over" incorporating remaining feedback into the code.

dcz-self commented:
On Tue, 27 Feb 2024 20:19:32 -0800
Amit Levy ***@***.***> wrote:

> This seems worth reviving to me (I have the hardware this was designed for, so I'm biased). This seems to have been very close and the dependent PRs on the surface seem fairly easy to approve other than the HIL discussion.
> 
> If @dcz-self has moved on, I'm willing to "take over" incorporating remaining feedback into the code.
> 

I'm ready to be part of the discussion, but I decided that I need to remove Tock from the dependency path of my project, the reason being that it's missing required functionality that is rather complex in comparison to less powerful OS, and reviewing progresses super slow even if I write it myself.

I'll be watching this to see if things improve.

bradjc commented:
Looking over this again, I think I'm arriving at the same conclusion that the discussion seemed to arrive at before. Basically, adding a new HIL that is similar to the flash HIL essentially introduces a split in the codebase, which leads to confusion and an inability to layer various components. If we are going to change what we have now, we really need to look at the entire nonvolatile storage stack holistically, how the layers fit together, and specify which interfaces every existing capsule should use. And then update those capsules at the same time.

alevy commented:
Closing without prejudice for lack of current momentum.

alevy commented:
Closing without prejudice for lack of current momentum.
